Add City Forecast scene daily forecast collection view

vandermesis opened the pull request.

vandermesis committed:
CollectionView lesson with Patryk

Changes:
	Setup of collectionview in CityForecastController
	Add CollectionView extension
	Setup of collectionview xib
	Tidy up extensions files

Fix:
	Fix CitiListInteractor fetching weather when user has no cities saved

vandermesis committed:
Modify CityForecast scene to display hourly and daily forecast for city

Changes:
	Add CityDailyForecast model
	Add CityDailyForecastListDisplayable model
	Modify CityForecastAPIResponse to get daily forecast
	Modify scene cells to show proper data
	Modify CityForecast scene methods to manage hourly and daily forecast data
	Rename CityForecast model
	Rename CityHourlyForecastListDisplayable model
	Modify DateFormatterHelper to properly show days and hours in lists
	Fix unit tests to handle all changes and cover new functionality

vandermesis committed:
Move CityForecast scene collectionview to first tableview cell

vandermesis committed:
Fix #23 Favourites not filtered when reenter scene

Fixed:
	Set search bar default text to empty string

vandermesis committed:
Change tableviews text font size

SredzinskiPatryk commented on CleanWeather/Extensions/NSObjectExtension.swift:
```diff
@@ -0,0 +1,26 @@
+//
+//  NSObjectExtension.swift
```
this filename tells nothing about the point of creating this extension
I'd name it: NSObjectClassNameExtension

and I'd rename other extensions as well to have their main goal in the file name. ErrorUserFriendlyMessageExtension etc

SredzinskiPatryk replied:
+ u can use '+' char in between to split object from it's purpose

![image](https://user-images.githubusercontent.com/32341779/71880850-7556c200-3131-11ea-9ef5-feacb67d571e.png)

SredzinskiPatryk commented on CleanWeather/Extensions/TableViewControllerExtension.swift:
```diff
@@ -8,25 +8,8 @@
 
 import UIKit
```
rename file to UITableViewCellDequeueExtension

SredzinskiPatryk commented on CleanWeather/Models/CityDailyForecast.swift:
```diff
@@ -0,0 +1,21 @@
+    let coordinates: String
+    let dateTimestamp: Double
+    let maxTemperature: Double
+    let minTemperatyre: Double
```
minTemperatyre

'y'?

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastController.swift:
```diff
@@ -54,30 +56,51 @@ extension CityForecastController: CityForecastPresentable {
+extension CityForecastController: UITableViewDelegate {
+
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        if indexPath.row == 0 {
```
having all these checks for indexPath.row == 0 is always looking not too clear, when you'll have to create more cells like that. 1 Above all, 2nd that can be or cannot be displayed depeneing on something, last cell different from datasource it gets complicated, so it's always better to make private extensions to handle that can have for example:

numberOfDailyForecastCells
numberOfHeaderCells
indexForDailyForecastCellAtIndexPath
isHeaderCell
isForecastCell

etc. think about it and try to make is more simple.
+ You can extract 3 lines in each dequeue, so u can have: buildHeaderCell and buildForecastCell

sometimes it's also good to move this logic to 'cellManager' class that you can have injected controller
and you call cellmanager in tableview datasource and delegate methods or maybe move whole datasource and delegate there?

test it and decide on your own how do you like it the most.
It's not bad right now, but you can do some experiments :)

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastPresenter.swift:
```diff
@@ -22,8 +22,12 @@ extension CityForecastPresenterImpl: CityForecastPresenter {
-        controller?.displayCityForecast(displayable)
+    func presentCityForecastList(_ forecast: CityForecast) {
+        let hourlyForecast = forecast.0
+        let dailyForecast = forecast.1
```
never use touples. Theey just look bad

how do I know what is .0 what is .1?
better to have structure:

forecast.hourly
forecast.daily

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastTableViewCell/CityForecastTableViewCellForCollectionView.swift:
```diff
@@ -0,0 +1,46 @@
+
+import UIKit
+
+class CityForecastTableViewCellForCollectionView: UITableViewCell {
```
CityForecastTableViewCellForCollectionView - wrong

CityForecastCollctionViewTableViiewCell - better
tableviewcell at the end tells you what it exactly is
name before tableviewcell - tells you what it's doing/having

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastWorker.swift:
```diff
@@ -25,16 +26,23 @@ final class CityForecastWorkerImpl {
+                                                                                       minTemperatyre: $0.temperatureLow,
+                                                                                       precipProbability: $0.precipProbability,
+                                                                                       icon: $0.icon)}
+                let cityForecast = (cityHourlyForecast, cityDailyForecast)
```
no tuples

SredzinskiPatryk commented on CleanWeatherTests/Mocks/Mock+CityForecastApiResponse.swift:
```diff
@@ -10,5 +10,5 @@
 extension Mock {
 
-    static let cityForecastApiResponse = CityForecastAPIResponse(hourly: CleanWeather.CityForecastAPIResponse.HourlyForecast(data: [CleanWeather.CityForecastAPIResponse.HourlyForecast.ForecastData(time: 1576713600.0, icon: "partly-cloudy-night", precipProbability: 0.0, temperature: 5.32), CleanWeather.CityForecastAPIResponse.HourlyForecast.ForecastData(time: 1576717200.0, icon: "partly-cloudy-nig ...
+    static let cityForecastApiResponse = CityForecastAPIResponse(hourly: CleanWeather.HourlyForecast(data: [CleanWeather.HourlyForecast.ForecastWeather(time: 1578247200.0, icon: "partly-cloudy-night", precipProbability: 0.04, temperature: -0.56), CleanWeather.HourlyForecast.ForecastWeather(time: 1578250800.0, icon: "partly-cloudy-night", precipProbability: 0.03, temperature: -0.94), CleanWeather. ...
```
add enters to make it mor clean

vandermesis committed:
Change custom extensions files naming convention

vandermesis committed:
Modify CityForecast tuple

Modified:
	Make CityForecast struct
	Remove CityForecast typealias

Fix:
	Fix unit tests after modifications
	Fix spelling in CityDailyForecast property

vandermesis committed:
Fix unit tests after modifications

vandermesis committed:
Change name of CityForecastCollectionViewTableViewCell

Changes:
	Renamed CityForecastTableViewCellForCollectionView
	Tidy up Mock+CityForecastApiResponse

vandermesis committed:
Modify CityForecast tableview cells setup

Modified:
	Add CellManager to manage cells setup
	Add CellManagerItems for different cells
	Inject CellManager to CityForecastController
	Some naming changes

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastController.swift:
```diff
@@ -10,22 +10,24 @@ import UIKit
-    private var cityForecastDataSource = [CityForecastListDisplayable]()
-    
-    init(interactor: CityForecastInteractor) {
+    private let cellManager: CellManager
```
Change cell manager name to have city forecast inside

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastController.swift:
```diff
@@ -54,30 +56,20 @@ extension CityForecastController: CityForecastPresentable {
+        forecastTableView.register(cellType: CityForecastTableViewCell.self)
+        forecastTableView.register(cellType: CityForecastCollectionViewTableViewCell.self)
+        forecastTableView.dataSource = cellManager
+        forecastTableView.separatorStyle = .none
```
You can set a separator style in Xib

SredzinskiPatryk commented on CleanWeather/Scenes/CityForecast/CityForecastTableViewCell/CellManager.swift:
```diff
@@ -0,0 +1,67 @@
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let item = items[indexPath.section]
+        switch item.type {
```
In most of cases if you have switch case you should put each cases into separate methods

vandermesis committed:
Change CellManager naming convention

Changes:
	Rename CellManager and it's components

Additional changes:
	Move CityForecast tableview separator setup to xib

vandermesis committed:
Refactor CellManager tableview cellForRow method

Changes:
	Move setup of different cells to separate methods

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Conflicts:
	CleanWeather.xcodeproj/project.pbxproj
